### PR TITLE
Session: don't return undefined if a response is required

### PR DESCRIPTION
**Pull request by andy-ms**
Sequel to #16773
This doesn't yet fix `SignatureHelp`, since that doesn't return an array. I'm not sure if VSCode is depending on us sending back an "error" response if the user isn't currently at a signature.

### CODE REVIEW

File: src/server/session.ts [hunk 1]
```diff
@@ -606,7 +606,7 @@ namespace ts.server {
 
             const definitions = project.getLanguageService().getDefinitionAtPosition(file, position);
             if (!definitions) {
-                return undefined;
+                return [];
```
Comment (rbuckton) [inline src/server/session.ts]: Are these results mutable? Can we use a shared `emptyArray` construct?

File: src/server/session.ts [hunk 2]
```diff
@@ -995,7 +995,7 @@ namespace ts.server {
             return args.position !== undefined ? args.position : scriptInfo.lineOffsetToPosition(args.line, args.offset);
         }
 
-        private getFileAndProject(args: protocol.FileRequestArgs, errorOnMissingProject = true) {
+        private getFileAndProject(args: protocol.FileRequestArgs, errorOnMissingProject = true): { file: NormalizedPath, project: Project } {
```
Comment (rbuckton) [inline src/server/session.ts]: Do we need the return type defined here, since it is correctly inferred?

**Comment by rbuckton** [comment]
Looks like the build is failing. Can you take a look?

**Comment by andy-ms** [comment]
Fix is #17685, will merge once that's in

**Comment by andy-ms** [comment]
@rbuckton Build is passing now, good to go?

File: src/server/session.ts [hunk 3]
```diff
@@ -813,7 +815,7 @@ namespace ts.server {
                 if (!renameInfo.canRename) {
                     return {
                         info: renameInfo,
-                        locs: []
+                        locs: emptyArray
```
Comment (amcasey) [inline src/server/session.ts]: Missed one on line 812?

File: src/server/session.ts [hunk 4]
```diff
@@ -909,7 +911,7 @@ namespace ts.server {
             if (simplifiedResult) {
                 const nameInfo = defaultProject.getLanguageService().getQuickInfoAtPosition(file, position);
                 if (!nameInfo) {
-                    return undefined;
```
Comment (amcasey) [inline src/server/session.ts]: Might break deserialization.

File: src/server/session.ts [hunk 5]
```diff
@@ -1172,7 +1174,7 @@ namespace ts.server {
 
             const completions = project.getLanguageService().getCompletionsAtPosition(file, position);
             if (!completions) {
-                return undefined;
```
Comment (amcasey) [inline src/server/session.ts]: Might break deserialization.

File: src/server/session.ts [hunk 6]
```diff
@@ -615,7 +617,7 @@ namespace ts.server {
             }
         }
 
-        private getTypeDefinition(args: protocol.FileLocationRequestArgs): protocol.FileSpan[] {
+        private getTypeDefinition(args: protocol.FileLocationRequestArgs): ReadonlyArray<protocol.FileSpan> {
```
Comment (amcasey) [inline src/server/session.ts]: Line 627?

File: src/server/session.ts [hunk 7]
```diff
@@ -687,25 +689,25 @@ namespace ts.server {
...
+        private getSemanticDiagnosticsSync(args: protocol.SemanticDiagnosticsSyncRequestArgs): ReadonlyArray<protocol.Diagnostic> | ReadonlyArray<protocol.DiagnosticWithLinePosition> {
             const { configFile, project } = this.getConfigFileAndProject(args);
             if (configFile) {
                 return this.getConfigFileDiagnostics(configFile, project, args.includeLinePosition);
             }
             return this.getDiagnosticsWorker(args, /*isSemantic*/ true, (project, file) => project.getLanguageService().getSemanticDiagnostics(file), args.includeLinePosition);
         }
 
-        private getDocumentHighlights(args: protocol.DocumentHighlightsRequestArgs, simplifiedResult: boolean): protocol.DocumentHighlightsItem[] | DocumentHighlights[] {
+        private getDocumentHighlights(args: protocol.DocumentHighlightsRequestArgs, simplifiedResult: boolean): ReadonlyArray<protocol.DocumentHighlightsItem> | ReadonlyArray<DocumentHighlights> {
```
Comment (amcasey) [inline src/server/session.ts]: Line 717?

**Comment by amcasey** [comment]
@andy-ms It might be safest to revert this (and hold off on https://github.com/Microsoft/TypeScript/pull/17727) until we further discuss the implications for clients. (cc: @minestarks, @rbuckton)

**Comment by andy-ms** [comment]
Might be better to wait for `release-2.5` and remove this from that, but not from `master`.

**Comment by amcasey** [comment]
I think `release-2.5` exists now.  The plan is to revert this there and submit https://github.com/Microsoft/TypeScript/pull/17728 to `master` (and close https://github.com/Microsoft/TypeScript/pull/17727)?

**Pull request merged**